handler/src/main/java/io/netty/handler/ssl/Java9SslUtils.java

```diff
@@ -0,0 +1,186 @@
+                final Class<?> biFunctionClass = biFunction;
+                selector = AccessController.doPrivileged(new PrivilegedExceptionAction<Method>() {
+                    @Override
+                    public Method run() throws Exception {
+                        return SSLEngine.class.getMethod("setHandshakeApplicationProtocolSelector", biFunctionClass);
+                    }
+                });
+            } catch (Throwable t) {
+                log.error("Unable to initialize Java9SslUtils, but the detected javaVersion was: {}", version, t);
+                getter = null;
+                setter = null;
+                biFunction = null;
+                selector = null;
+            }
+        }
+
+        GET_APPLICATION_PROTOCOL = getter;
+        SET_APPLICATION_PROTOCOL = setter;
+        BI_FUNCTION_CLASS = biFunction;
+        SET_HANDSHAKE_APPLICATION_PROTOCOL_SELECTOR = selector;
+    }
+
+    private Java9SslUtils() {
+    }
+
+    static boolean supportsAlpn() {
+        return GET_APPLICATION_PROTOCOL != null;
+    }
+
+    static void configureAlpn(JdkSslEngine engine, JdkApplicationProtocolNegotiator applicationNegotiator) {
+        List<String> supportedProtocols = applicationNegotiator.protocols();
+
+        SSLEngine wrapped = engine.getWrappedEngine();
+        SSLParameters params = wrapped.getSSLParameters();
+        setApplicationProtocols(params, supportedProtocols);
+        wrapped.setSSLParameters(params);
+
+        if (!wrapped.getUseClientMode()) {
+            installSelector(engine, applicationNegotiator);
+        }
+    }
+
+    static String getApplicationProtocol(SSLEngine sslEngine) {
+        try {
+            return (String) GET_APPLICATION_PROTOCOL.invoke(sslEngine);
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static void setApplicationProtocols(SSLParameters parameters, List<String> supportedProtocols) {
+        String[] protocolArray = supportedProtocols.toArray(EmptyArrays.EMPTY_STRINGS);
+        try {
+            SET_APPLICATION_PROTOCOL.invoke(parameters, new Object[]{ protocolArray });
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static void installSelector(final JdkSslEngine sslEngine,
+                                        final JdkApplicationProtocolNegotiator applicationProtocolNegotiator) {
+        checkNotNull(applicationProtocolNegotiator, "applicationProtocolNegotiator");
+
+        Object biFunction = AccessController.doPrivileged(new PrivilegedAction<Object>() {
+            @Override
+            public Object run() {
+                return Proxy.newProxyInstance(Java9SslUtils.class.getClassLoader(), new Class[]{ BI_FUNCTION_CLASS },
+                        new AlpnSelector(sslEngine, applicationProtocolNegotiator));
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Avoid instantiation through private constructors from outside of the constructors class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AAccessorClassGeneration)